module/iso_fortran_env.f90

```diff
@@ -19,20 +19,110 @@
 
 module iso_fortran_env
 
-  integer, parameter :: atomic_int_kind = 8
-  integer, parameter :: atomic_logical_kind = 8
+  implicit none
+
+  integer, parameter :: atomic_int_kind = selected_int_kind(18)
+  integer, parameter :: atomic_logical_kind = atomic_int_kind
+
+  ! TODO: Use PACK([x],test) in place of the array constructor idiom
+  ! [(x, integer::j=1,COUNT([test]))] below once PACK() can be folded.
+
+  integer, parameter, private :: &
+    selectedASCII = selected_char_kind('ASCII'), &
+    selectedUCS_2 = selected_char_kind('UCS-2'), &
+    selectedUnicode = selected_char_kind('ISO_10646')
+  integer, parameter :: character_kinds(*) = [ &
+    [(selectedASCII, integer :: j=1, count([selectedASCII >= 0]))], &
+    [(selectedUCS_2, integer :: j=1, count([selectedUCS_2 >= 0]))], &
+    [(selectedUnicode, integer :: j=1, count([selectedUnicode >= 0]))]]
+
+  integer, parameter, private :: &
+    selectedInt8 = selected_int_kind(2), &
+    selectedInt16 = selected_int_kind(4), &
+    selectedInt32 = selected_int_kind(9), &
+    selectedInt64 = selected_int_kind(18),&
+    selectedInt128 = selected_int_kind(38), &
+    safeInt8 = merge(selectedInt8, selected_int_kind(0), &
+                     selectedInt8 >= 0), &
+    safeInt16 = merge(selectedInt16, selected_int_kind(0), &
+                      selectedInt16 >= 0), &
+    safeInt32 = merge(selectedInt32, selected_int_kind(0), &
+                      selectedInt32 >= 0), &
+    safeInt64 = merge(selectedInt64, selected_int_kind(0), &
+                      selectedInt64 >= 0), &
+    safeInt128 = merge(selectedInt128, selected_int_kind(0), &
+                       selectedInt128 >= 0)
+  integer, parameter :: &
+    int8 = merge(selectedInt8, merge(-2, -1, selectedInt8 >= 0), &
+                 digits(int(0,kind=safeInt8)) == 7), &
+    int16 = merge(selectedInt16, merge(-2, -1, selectedInt16 >= 0), &
+                  digits(int(0,kind=safeInt16)) == 15), &
+    int32 = merge(selectedInt32, merge(-2, -1, selectedInt32 >= 0), &
+                  digits(int(0,kind=safeInt32)) == 31), &
+    int64 = merge(selectedInt64, merge(-2, -1, selectedInt64 >= 0), &
+                  digits(int(0,kind=safeInt64)) == 63), &
+    int128 = merge(selectedInt128, merge(-2, -1, selectedInt128 >= 0), &
+                   digits(int(0,kind=safeInt128)) == 127)
+  integer, parameter :: integer_kinds(*) = [ &
+    [(int8, integer :: j=1, count([int8==selectedInt8]))], &
```

Comment:
Why not `count([int8 >= 0])`? That's how it's done for `character_kinds` and `logical_kinds` but not `integer_kinds` and `real_kinds`.

Reply:
Ok.

Reply:
Thanks for pointing this out -- both `integer_kinds` and `real_kinds` can include kinds other than the standard ones, since they're meant to include all of the kinds supported on the "processor".  Fixing this with a complicated search loop written as an array constructor exposed a bug in array constructor expression analysis, and it's good that it was caught now.